### PR TITLE
Added simple endpoint for PATCH for course_runs api to allow the modi…

### PR DESCRIPTION
**Pull request by davec-edx**
Addition of a PATCH verb to the API

ECOM-7024

### CODE REVIEW

**Review by clintonb — CHANGES_REQUESTED** [review]
Permissions need to be fixed. We don't want any authenticated user to be able to modify course runs.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 1]
```diff
@@ -103,6 +103,29 @@ def test_get_include_unpublished_programs(self):
                 self.serialize_course_run(self.course_run, extra_context={'include_unpublished_programs': True})
             )
 
+    def test_partial_update(self):
+        """ Verify the endpoint supports partially updating a course_run's fields. """
```
Comment (MatthewPiatetsky) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: this works, but what if we used the CourseRun factory and updated all the fields, then checked that all of them got updated. If all the fields including m2m actually work, using the factory could be a simple way to ensure that new fields also get updated without having to update the test

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 2]
```diff
@@ -103,6 +103,29 @@ def test_get_include_unpublished_programs(self):
...
+        }
+
+        # Update this course_run with the new info
+        response = self.client.patch(url, data, format='json')
+        self.assertEqual(response.status_code, 200)
+
+        # re-read and make sure we have the newly calculated effort levels
+        updated_course_run = CourseRun.objects.get(key=self.course_run.key)
+
+        assert updated_course_run.max_effort == max_effort
```
Comment (MatthewPiatetsky) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: if you go with the factory you could probably have a loop of assert statements using the original_course_run.__dict__.items() and updated_course_run.__dict__.items() or something like that

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 3]
```diff
@@ -23,7 +23,7 @@
 class CourseRunViewSetTests(SerializationMixin, ElasticsearchTestMixin, APITestCase):
     def setUp(self):
         super(CourseRunViewSetTests, self).setUp()
-        self.user = UserFactory(is_staff=True, is_superuser=True)
+        self.user = UserFactory(is_staff=False, is_superuser=True)
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: This does not need to be changed.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 4]
```diff
@@ -103,6 +103,47 @@ def test_get_include_unpublished_programs(self):
                 self.serialize_course_run(self.course_run, extra_context={'include_unpublished_programs': True})
             )
 
+    def test_partial_update(self):
+        """ Verify the endpoint supports partially updating a course_run's fields, provided user has permission. """
+        url = reverse('api:v1:course_run-detail', kwargs={'key': self.course_run.key})
+        self.request.user.is_staff = True
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: This has no effect on the test.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 5]
```diff
@@ -103,6 +103,47 @@ def test_get_include_unpublished_programs(self):
...
+            'max_effort': max_effort,
+            'min_effort': min_effort,
+        }
+
+        # Update this course_run with the new info
+        response = self.client.patch(url, data, format='json')
+        self.assertEqual(response.status_code, 200)
+
+        # re-read and make sure we have the newly calculated effort levels
+        updated_course_run = CourseRun.objects.get(key=self.course_run.key)
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: Use [`refresh_from_db`](https://docs.djangoproject.com/en/1.10/ref/models/instances/#django.db.models.Model.refresh_from_db).

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 6]
```diff
@@ -103,6 +103,47 @@ def test_get_include_unpublished_programs(self):
                 self.serialize_course_run(self.course_run, extra_context={'include_unpublished_programs': True})
             )
 
+    def test_partial_update(self):
+        """ Verify the endpoint supports partially updating a course_run's fields, provided user has permission. """
+        url = reverse('api:v1:course_run-detail', kwargs={'key': self.course_run.key})
+        self.request.user.is_staff = True
+
+        # Get a course_run, and calculate new min_ and max_effort
+        original_course_run = CourseRun.objects.get(key=self.course_run.key)
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: There is no need to reload this object from the database. Your test values don't need to depend on the original values.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 7]
```diff
@@ -103,6 +103,47 @@ def test_get_include_unpublished_programs(self):
...
+        max_effort = original_course_run.max_effort + 867
+        min_effort = original_course_run.min_effort + 5309
+        data = {
+            'max_effort': max_effort,
+            'min_effort': min_effort,
+        }
+
+        # Update this course_run with the new info
+        response = self.client.patch(url, data, format='json')
+        self.assertEqual(response.status_code, 200)
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: `assert response.status_code == 200`

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 8]
```diff
@@ -127,6 +128,13 @@ def list(self, request, *args, **kwargs):
         """
         return super(CourseRunViewSet, self).list(request, *args, **kwargs)
 
+    def partial_update(self, request, *args, **kwargs):
+        """ Update one, or more, fields for a course run. """
+        if self.request.user.is_authenticated() and self.request.user.is_staff:
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/course_runs.py]: This is not how we handle permissions. See http://www.django-rest-framework.org/api-guide/permissions/.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 9]
```diff
@@ -103,6 +103,47 @@ def test_get_include_unpublished_programs(self):
...
+        updated_course_run = CourseRun.objects.get(key=self.course_run.key)
+
+        assert updated_course_run.max_effort == max_effort
+        assert updated_course_run.min_effort == min_effort
+
+    def test_partial_update_bad_permission(self):
+        """ Verify the endpoint supports partially updating will fail if user doesn't have permission. """
+        url = reverse('api:v1:course_run-detail', kwargs={'key': self.course_run.key})
+
+        # Get a course_run, and calculate new min_ and max_effort
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: This test should be much simpler. There is no need to worry about the data being sent to the API because we should not get to that level.

```python
response = self.client.patch(url,{}, format='json')
assert response.status_code == 403
```

**Review by clintonb — CHANGES_REQUESTED** [review]

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py (hunk 3 above)
Comment (davec-edx) [reply course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: Everything was being tested with is_staff set, where most don't need anything more than being authenticated.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py (hunk 4 above)
Comment (davec-edx) [reply course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: In conjunction with not setting it for other tests, it needs to be set for this test (and unset for the corresponding negative test)

**Comment by davec-edx** [comment]
@clintonb Please re-review -- thanks

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py (hunk 2 above)
Comment (davec-edx) [reply course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: Yes, I thought about that, but thought it would be overkill, I just chose a couple fields at random.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py (hunk 3 above)
Comment (clintonb) [reply course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: Superuser trumps staff. Given that all superusers are staff, this change is not necessary.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py (hunk 4 above)
Comment (clintonb) [reply course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: If you revert the change to `self.user` this line can go away.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py (hunk 6 above)
Comment (clintonb) [reply course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: Please clean this up. There is no need to load the object from the DB. Doing so is unnecessarily confusing.

```python
expected_max_effort = 1000
expected_min_effort = 100
data = {
  'max_effort': expected_max_effort,
  'min_effort': expected_min_effort,
}

response = self.client.patch(url, data, format='json')
assert response.status_code == 200

self.course_run.refresh_from_db()
assert self.course_run.max_effort == expected_max_effort
assert self.course_run.min_effort == expected_min_effort
```

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 10]
```diff
@@ -103,6 +103,39 @@ def test_get_include_unpublished_programs(self):
...
+        updated_course_run = CourseRun.objects.get(key=self.course_run.key)
+
+        assert updated_course_run.max_effort == max_effort
+        assert updated_course_run.min_effort == min_effort
+
+    def test_partial_update_bad_permission(self):
+        """ Verify the endpoint supports partially updating will fail if user doesn't have permission. """
+        url = reverse('api:v1:course_run-detail', kwargs={'key': self.course_run.key})
+
+        # Update this course_run with the new info, NOTE this should fail with a 404
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: This is not valid.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 11]
```diff
@@ -103,6 +103,39 @@ def test_get_include_unpublished_programs(self):
...
+        response = self.client.patch(url, {}, format='json')
+        assert response.status_code != 200
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: `assert response.status_code == 403`

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 12]
```diff
@@ -14,14 +16,14 @@
...
-class CourseRunViewSet(PartnerMixin, viewsets.ReadOnlyModelViewSet):
+class CourseRunViewSet(PartnerMixin, viewsets.ModelViewSet):
     """ CourseRun resource. """
     filter_backends = (DjangoFilterBackend, OrderingFilter)
     filter_class = filters.CourseRunFilter
     lookup_field = 'key'
     lookup_value_regex = COURSE_RUN_ID_REGEX
     ordering_fields = ('start',)
-    permission_classes = (IsAuthenticated,)
+    permission_classes = (IsAuthenticated, )
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/course_runs.py]: Revert

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 13]
```diff
@@ -127,6 +129,11 @@ def list(self, request, *args, **kwargs):
         """
         return super(CourseRunViewSet, self).list(request, *args, **kwargs)
 
+    @method_decorator(staff_member_required)
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/course_runs.py]: This is not what we discussed yesterday. You should be using [`DjangoModelPermissions`](http://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions).

**Review by clintonb — CHANGES_REQUESTED** [review]

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 14]
```diff
@@ -14,14 +14,14 @@
...
-class CourseRunViewSet(PartnerMixin, viewsets.ReadOnlyModelViewSet):
+class CourseRunViewSet(PartnerMixin, viewsets.ModelViewSet):
     """ CourseRun resource. """
     filter_backends = (DjangoFilterBackend, OrderingFilter)
     filter_class = filters.CourseRunFilter
     lookup_field = 'key'
     lookup_value_regex = COURSE_RUN_ID_REGEX
     ordering_fields = ('start',)
-    permission_classes = (IsAuthenticated,)
+    permission_classes = (DRYPermissions,)
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/course_runs.py]: `permission_classes = (IsAuthenticated, DjangoModelPermissions)`

We don't need instance-specific permissions, and we always want users to be authenticated. Use `DjangoModelPermissions`.

File: course_discovery/apps/course_metadata/models.py [hunk 15]
```diff
@@ -338,7 +339,7 @@ def reindex_course_runs(self):
             index.update_object(course_run)
 
 
-class CourseRun(TimeStampedModel):
+class CourseRun(ModelPermissionsMixin, TimeStampedModel):
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: This is not necessary.

**Review by clintonb — CHANGES_REQUESTED** [review]

**Review by clintonb — APPROVED** [review]
